### ISSUE TITLE
Apple clang version misdetected

### ISSUE THREAD

**Issue by Ionic** [mesonbuild/meson#6253]
On (Mac) OS X/macOS, Apple patches the clang version as shipped with Xcode and inserts a special version (mostly based upon the Xcode major version).

For instance, Xcode 6.2 ships with a clang that identifies itself as "6.0 (clang-600.0.57) (based on LLVM 3.5svn)".

meson extracts the version number more or less correctly, but that leads to a lot of weird effects later on.

For instance, it later [uses this version when adding a compiler flag](https://github.com/mesonbuild/meson/blob/adcd70861d14ae591130cc1173093e5c83513867/mesonbuild/compilers/mixins/clang.py#L66). This flag is then added, since the parsed version number (`6.0`) is higher than the version number compared against (`3.6.0`) and... stuff blows up afterwards, because the compiler is actually a 3.4.0-based one (or actually somewhere between 3.4.0 and 3.5.0, sadly).

This is just one example of things breaking, there are probably more.

Fixing that issue will be rather difficult, though.

To *correctly* detect the version number, I think that meson has to map the major Apple clang version (for instance, `600` in the example above) to an actual LLVM release (again, for instance `3.4.0` in the example above). The actual version number it reports, such as `6.0`, is utterly useless and shouldn't be exposed at all.

This requires maintaining a table of major version number to actual clang/LLVM release version number mappings that will have to be updated manually. Not sure if that's a great idea.

Note that this does **not** apply universally to clang on OS X/macOS. This issue affects Xcode-based clang compilers only.

**Comment by scivision** [comment]
Is this the area where the detection that's affecting this issue is being done?
https://github.com/mesonbuild/meson/commit/c756d06fbb54651c2ba19f7801885018c1c389df#diff-4f050ed89f9a7e15b09dd12b1e940a19R94

**Comment by Ionic** [comment]
Well... not really. It's a fine example of this issue being worked around, but not properly fixed in general.

What this commit did is to check against an Apple-clang version of 10, which roughly corresponds to a clang 6.0.1 version.

It's not a generic solution, though, since any and all such clang version checks would have to be wrapped in something like this, which might be doable within meson itself, but expecting that from consumers of meson would be just cruel (and, frankly, almost no project would get it right).

**Comment by scivision** [comment]
Yes that's what I meant, that this was commit where the decision wasn't being done with enough fidelity

**Comment by Ionic** [comment]
Urgh... I just noticed that master got changes such as ff4a17dbef08a1d8afd075f57dbab0f5c76951ab and 0c22798b1ad4678abb205280060175678a790c4a.

That's... a step in a direction I don't like. It does abstract the clang compiler and adds some more specific implementation like the Apple/Xcode-based one, but it keeps the version numbering around, so keeps this issue alive.

I've patched 0.52.0 first and tested it successfully against a failing project, but then noticed the rewrite.

I've rebased my changes against that, see https://github.com/Ionic/meson/commit/bea64af3671b2535760b65127a120d50c11b0a5e. One takeaway is that I haven't actually tested these changes yet. I'm also seeking guidance on my approach and the way I implemented it - I'm not very familiar with Python.

**Comment by someuser12** [comment]
Further examples of misdetections of Apple clang:

- Under Mac OS 10.13.6, `/usr/bin/clang --version` reports `Apple LLVM version 9.0.0 (clang-900.0.38)` but meson claims `C compiler for the host machine: /usr/bin/clang (clang 10.0.0 "Apple LLVM version 10.0.0 (clang-1000.10.44.4)")` (while configuring gdk-pixbuf2).

- Under Mac OS 10.11.6, `/usr/bin/clang --version` reports `Apple LLVM version 7.0.2 (clang-700.1.81)` but meson claims `C compiler for the host machine: /usr/bin/clang (clang 8.0.0 "Apple LLVM version 8.0.0 (clang-800.0.42.1)")` (while configuring gdk-pixbuf2).

**Comment by Ionic** [comment]
This is probably not a misdetection but a machine misconfiguration.

`meson` takes clang's version output literally, so you *do* have to different set of compilers installed.

Your system compiler is older which means that you have likely upgraded Xcode itself correctly, but didn't also install/upgrade the new command line tools.

Do that and both versions should match. Also keep in mind to do this on every further Xcode upgrade. It doesn't do that for you automatically. :)

**Comment by someuser12** [comment]
I'm a bit puzzled here, since I can't find where this different set of compilers would be installed:
```
$ sw_vers
ProductName:    Mac OS X
ProductVersion: 10.13.6
BuildVersion:   17G11023
$ find /Applications/Xcode.app/ -name \*clang -type f
/Applications/Xcode.app//Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang
$ /Applications/Xcode.app//Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang --version
Apple LLVM version 9.0.0 (clang-900.0.38)
Target: x86_64-apple-darwin17.7.0
Thread model: posix
InstalledDir: /Applications/Xcode.app//Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin
$ /usr/bin/clang --version
Apple LLVM version 9.0.0 (clang-900.0.38)
Target: x86_64-apple-darwin17.7.0
Thread model: posix
InstalledDir: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin
$
```

Also, I don't understand why meson reports that the C compiler is `/usr/bin/clang` if it actually runs another compiler (`/usr/bin/clang`'s literal output is version 9.0.0, so meson must be running another file). Is there a way to ask meson to tell precisely which binary it runs?

**Comment by someuser12** [comment]
To answer my own questions: this is because `/usr/bin/clang` was called (through MacPorts) with the environment variable `DEVELOPER_DIR` set to `/Library/Developer/CommandLineTools`, where the other set of compiler is installed.

So this has nothing to do with meson.

**Comment by dcbaker** [comment]
We can close this then?

**Comment by Ionic** [comment]
Not at all. :)

someuser12's issue was just unrelated to this. Sorry, I could have been more specific to begin with and give you pointers to the `/Library/Developer/CommandLineTools` dir to spare you the search.

**Comment by bonzini** [comment]
There are a few remaining issues:

* Only the C and C++ Apple compiler classes pull in AppleCompilerMixin, Objective-C and Objective-C++ don't
* While OpenMP and prelinking do have some overrides for Apple compilers, there are still some missing cases
  - [ ]  -fpch-instantiate-templates (>=11)
  - [ ] _LIBCPP_HARDENING_MODE_DEBUG (>=18)
  - [ ] -flto-jobs (4.0.0, should be fine)
  - [ ] -Werror=ignored-optimization-argument (3.6.0, should be fine)
